### PR TITLE
Update Adam for iDeep 2.0

### PR DESCRIPTION
**Pull request by mingxiaoh**
Thank you for creating a pull request!

Please double-check the following.

- Read [our contribution guide](https://docs.chainer.org/en/stable/contribution.html).
  - Does you code conform to our coding guidelines?
  - Did you write sufficient test code?
  - Did you write sufficient documentation?
- Also take a look at [our compatibility policy](https://docs.chainer.org/en/stable/compatibility.html).

### CODE REVIEW

File: chainer/optimizers/adam.py [hunk 1]
```diff
@@ -99,18 +110,30 @@ def update_core_cpu(self, param):
...
-        v += (1 - hp.beta2) * (grad * grad - v)
-
-        if hp.amsgrad:
-            vhat = self.state['vhat']
-            numpy.maximum(vhat, v, out=vhat)
+        if (isinstance(m, intel64.mdarray) and
+            isinstance(v, intel64.mdarray)):
+            m.inplace_axpby(1.0, 1.0 - hp.beta1, grad - m)
+            v.inplace_axpby(1.0, 1.0 - hp.beta2, grad*grad - v)
+            if hp.amsgrad:
```
Comment (kmaehashi) [inline chainer/optimizers/adam.py]: Let's align the code style with the non-ideep case.

```py
```
if hp.amsgrad:
    vhat = self.state['vhat']
    numpy.maximum(vhat, v, out=vhat)
else:
    vhat = v
param.data.inplace_axpby(1.0 - hp.weight_decay_rate, -hp.eta,
                         self.lr * m / (numpy.sqrt(vhat) + hp.eps))
```

**Comment by kmaehashi** [comment]
* Could you add test case?
* What does "modified by recomendation [provided by Ikei]" mean?

**Comment by mingxiaoh** [comment]
- Could you add test case?
[mingxiao]: will provide it later
What does "modified by recomendation [provided by Ikei]" mean?
[mingxiao]: this PR was originally provided by Ikei.

**Comment by mingxiaoh** [comment]
@kmaehashi  I am trying to add corresponding test case, however, it seems that no specific unit test case for AdamRule originally, correct? If no, would you please tell me which test file and function is for AdamRule? Thanks in advance.

**Comment by kmaehashi** [comment]
Ah, OK, it seems ideep case is already covered. No need to add test.
https://github.com/chainer/chainer/blob/v5.0.0b2/tests/chainer_tests/optimizers_tests/test_optimizers_by_linear_model.py#L103-L112

Jenkins, test this please.

**Comment by mingxiaoh** [comment]
@kmaehashi when will the code be merged?

**Comment by kmaehashi** [comment]
LGTM! Test failures are not related to this PR.
Will merge after #4933.

**Comment by mingxiaoh** [comment]
@niboshi when will this PR be merged?

**Comment by mingxiaoh** [comment]
@kmaehashi @niboshi when will this PR be merged?

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
LGTM!

**Pull request merged**